File: Python/easy/1071. Greatest Common Divisor of Strings.py
"""
For two strings s and t, we say "t divides s" if and only if s = t + ... + t  (t concatenated with itself 1 or more times)

Given two strings str1 and str2, return the largest string x such that x divides both str1 and str2.

 

Example 1:

Input: str1 = "ABCABC", str2 = "ABC"
Output: "ABC"
Example 2:

Input: str1 = "ABABAB", str2 = "ABAB"
Output: "AB"
Example 3:

Input: str1 = "LEET", str2 = "CODE"
Output: ""
Example 4:

Input: str1 = "ABCDEF", str2 = "ABC"
Output: ""
 

Constraints:

1 <= str1.length <= 1000
1 <= str2.length <= 1000
str1 and str2 consist of English uppercase letters.
"""

class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        if len(str1) > len(str2):
            return self.gcdOfStrings(str2, str1)
        result = ""
        if str1 * (len(str2) // len(str1)) == str2:
            return str1
        for i in range(len(str1) // 2, 0, -1):
            if str1[:i] * (len(str1) // i) == str1 and str1[:i] * (len(str2) // i) == str2:
                return str1[:i]
        return ""

File: Java/easy/1071. Greatest Common Divisor of Strings.java
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        if (str1.length() > str2.length()) {
            return gcdOfStrings(str2, str1);
        }
        if (str1.repeat(str2.length() / str1.length()).equals(str2)) {
            return str1;
        }
        for (int i = str1.length() / 2; i > 0; i--) {
            String result = str1.substring(0, i);
            if (result.repeat(str1.length() / i).equals(str1) & result.repeat(str2.length() / i).equals(str2)) {
                return result;
            } 
        }
        return "";
    }
}